spec/Classes.tex

```diff
@@ -739,37 +1063,348 @@ \subsection{The Compiler-Generated Constructor}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+\begin{chapeloutput}
+{x = 0.0, y = 0.0, z = 5.0}
+\end{chapeloutput}
+The second initializer does not contain an implicit call to the parent's
+initializer because it explicitly invokes another initializer.
+\end{chapelexample}
+
+\subsubsection{Calling Methods on Parent Classes}
+\label{Calling_Methods_on_Parent_Classes}
+\index{classes!initializers!initializer-inheritance!dynamic-this}
+
+Once \chpl{super.init()} returns, the dynamic type of \chpl{this} is the
+parent's type until the \chpl{complete} method is invoked (except when the
+child's fields are initialized and used). As a result, the parent's methods
+may be called and \chpl{this} may be passed to functions as the parent type.
+
+\begin{rationale}
+After \chpl{super.init()} returns the instance is in some
+partially-initialized, but valid, state. Allowing \chpl{this} to be treated as
+the parent allows for additional functionality and flexibility for users.
+\end{rationale}
+
+\begin{chapelexample}{dynamicThisInit.chpl}
+In the following code:
+\begin{chapel}
+class Parent {
+  var x, y : real;
+
+  proc foo() {
+    writeln("Parent.foo");
+  }
+}
+
+class Child : Parent {
+  var z : real;
+
+  proc init(x: real, y: real, z: real) {
+    super.init(x, y); // compiler-generated initializer
+    foo(); // Parent.foo()
+    this.z = z;
+    this.complete();
+    foo(); // Child.foo()
+  }
+
+  override proc foo() {
+    writeln("Child.foo");
+  }
+}
+
+var c = new Child(1.0, 2.0, 3.0);
+\end{chapel}
+\begin{chapelpost}
+writeln(c);
+\end{chapelpost}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+Once the parent's initializer is finished, the parent method
+\chpl{foo} may be called. After the \chpl{complete} method is invoked, a call to \chpl{foo}
+resolves to the Child's overriden implementation:
+\begin{chapelprintoutput}{}
+Parent.foo
+Child.foo
+{x = 1.0, y = 2.0, z = 3.0}
+\end{chapelprintoutput}
+\end{chapelexample}
+
+\subsubsection{The Compiler Generated Initializer for Inheriting Classes}
+\label{The_Compiler_Generated_Initializer_for_Inheriting_Classes}
+\index{classes!initializers!initializer-inheritance!compiler-generated}
+
+The compiler-generated initializer for inheriting classes will have arguments
+with default values and names based on the field declarations in the parent
+class. Formals for the parent type will be listed before formals for the
+child type.
+
+\begin{chapelexample}{compilerGeneratedInheritanceInit.chpl}
+In the following code:
+\begin{chapel}
+class Parent {
+  var x, y: real;
+}
+
+class Child : Parent {
+  var z : real;
+}
+
+var A = new Child();
+var B = new Child(1.0, 2.0, 3.0);
+var C = new Child(y=10.0);
+\end{chapel}
+\begin{chapelpost}
+writeln(A);
+writeln(B);
+writeln(C);
+\end{chapelpost}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+\begin{chapeloutput}
+{x = 0.0, y = 0.0, z = 0.0}
+{x = 1.0, y = 2.0, z = 3.0}
+{x = 0.0, y = 10.0, z = 0.0}
```

Comment:
I would either show this output, or include a comment on the declaration showing which fields the numbers in `var B` will go to